### PR TITLE
Change rpm and tpm to have lower default and allow for manual setting

### LINKED ISSUE #232: Manually set rate limits as argument to Prompter
For example, when using the SambaNova API with the litellm backend with curator, there is no way of getting the rate limit via the API. Allow the user to set RPM and TPM.

### PR DESCRIPTION
**Pull request by RyanMarten**
Fixes #232

### CODE REVIEW

**Comment by RyanMarten** [comment]
## Test defaults

Test claude-3-5-haiku (litellm backend with model that has rate limit headers)
```
python tests/simple_online.py --log-level DEBUG --model claude-3-5-haiku-20241022
```
```
2024-12-09 14:44:00,561 - bespokelabs.curator.request_processor.litellm_online_request_processor - INFO - Getting rate limits for model: claude-3-5-haiku-20241022
2024-12-09 14:44:01,168 - bespokelabs.curator.request_processor.litellm_online_request_processor - INFO - Rate limit headers: {'x-ratelimit-limit-requests': '4000', 'x-ratelimit-remaining-requests': '3999', 'x-ratelimit-limit-tokens': '480000', 'x-ratelimit-remaining-tokens': '472000', 'llm_provider-date': 'Mon, 09 Dec 2024 22:44:01 GMT', 'llm_provider-content-type': 'application/json', 'llm_provider-transfer-encoding': 'chunked', 'llm_provider-connection': 'keep-alive', 'llm_provider-anthropic-ratelimit-requests-limit': '4000', 'llm_provider-anthropic-ratelimit-requests-remaining': '3999', 'llm_provider-anthropic-ratelimit-requests-reset': '2024-12-09T22:44:00Z', 'llm_provider-anthropic-ratelimit-input-tokens-limit': '400000', 'llm_provider-anthropic-ratelimit-input-tokens-remaining': '400000', 'llm_provider-anthropic-ratelimit-input-tokens-reset': '2024-12-09T22:44:01Z', 'llm_provider-anthropic-ratelimit-output-tokens-limit': '80000', 'llm_provider-anthropic-ratelimit-output-tokens-remaining': '72000', 'llm_provider-anthropic-ratelimit-output-tokens-reset': '2024-12-09T22:44:06Z', 'llm_provider-anthropic-ratelimit-tokens-limit': '480000', 'llm_provider-anthropic-ratelimit-tokens-remaining': '472000', 'llm_provider-anthropic-ratelimit-tokens-reset': '2024-12-09T22:44:01Z', 'llm_provider-request-id': 'req_015o2Fwhh4Agkgy15CPr3H3B', 'llm_provider-via': '1.1 google', 'llm_provider-cf-cache-status': 'DYNAMIC', 'llm_provider-x-robots-tag': 'none', 'llm_provider-server': 'cloudflare', 'llm_provider-cf-ray': '8ef884700ac167a0-SJC', 'llm_provider-content-encoding': 'gzip', 'llm_provider-x-ratelimit-limit-requests': '4000', 'llm_provider-x-ratelimit-remaining-requests': '3999', 'llm_provider-x-ratelimit-limit-tokens': '480000', 'llm_provider-x-ratelimit-remaining-tokens': '472000'}
2024-12-09 14:44:01,168 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Automatically set max_requests_per_minute to 4000
2024-12-09 14:44:01,168 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Automatically set max_tokens_per_minute to 480000
```

Test gpt-4o-mini (openai backend)
```
python tests/simple_online.py --log-level DEBUG --model gpt-4o-mini

```
```
2024-12-09 14:46:03,500 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Automatically set max_requests_per_minute to 30000
2024-12-09 14:46:03,500 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Automatically set max_tokens_per_minute to 150000000
```

Test sambanova (litellm backend with a model that has no rate limit in headers)
```
python tests/simple_online.py --log-level DEBUG --model sambanova/Meta-Llama-3.1-8B-Instruct
```
```
2024-12-09 14:48:00,522 - bespokelabs.curator.request_processor.base_online_request_processor - WARNING - No manual max_requests_per_minute set, and automatic detection failed, using default value of 10
2024-12-09 14:48:00,522 - bespokelabs.curator.request_processor.base_online_request_processor - WARNING - No manual max_tokens_per_minute set, and automatic detection failed, using default value of 1000
```

**Comment by RyanMarten** [comment]
## Test manually setting

openai backend
```
python tests/simple_online.py --log-level DEBUG --model gpt-4o-mini --n-requests 1000 
```
```
python tests/simple_online.py --log-level DEBUG --model gpt-4o-mini --n-requests 1000 --max-requests-per-minute 60
```

litellm backend

purposely set too high
```
python tests/simple_online.py --log-level DEBUG --model sambanova/Meta-Llama-3.1-8B-Instruct --n-requests 1000 --max-requests-per-minute 10000 --max-tokens-per-minute 10000000000
```
start getting too many request errors
```
Request 587 failed with Exception litellm.RateLimitError: RateLimitError: SambanovaException - <html>
<head><title>429 Too Many Requests</title></head>
<body>
<center><h1>429 Too Many Requests</h1></center>
<hr><center>nginx</center>
</body>
</html>, attempts left 5
```

set to the right amount (QPM) 
```
python tests/simple_online.py --log-level DEBUG --model sambanova/Meta-Llama-3.1-8B-Instruct --n-requests 1000 --max-requests-per-minute 600 --max-tokens-per-minute 10000000000
```
No issues

File: src/bespokelabs/curator/request_processor/base_online_request_processor.py [hunk 1]
```diff
@@ -473,3 +479,28 @@ async def append_generic_response(self, data: GenericResponse, filename: str) ->
         async with aiofiles.open(filename, "a") as f:
             await f.write(json_string + "\n")
         logger.debug(f"Successfully appended response to {filename}")
+
+    def rate_limit_helper(self, rpm, tpm):
```
Comment (vutrung96) [inline src/bespokelabs/curator/request_processor/base_online_request_processor.py:483]: the way this abstraction is defined is confusing imo.

iiuc currently:
* rate_limit_helper (defined by base class) merges the manually-set rate limit, the rate limit from the header, and the fallback rate limit.
* get_rate_limits (called by child classes) gets the rate limit from header and call rate_limit_helper from base class

i think it's unclear from the method name what is happening, and also people who implement child class needs to remember to call rate_limit_helper method at the end which is kind of boiler-plate. so what about:

* get_header_based_rate_limits (abstract method that needs to be implemented by child classes) returns rate limits from headers
* get_rate_limits (defined by base class) does the merging from get_header_based_rate_limits, manually-set rate limit from __init__, and fallback default rate limits.

also, get_rate_limits is not defined by the base class as an abstract method that needs to be implemented, but i think it should be.

Comment (RyanMarten) [reply src/bespokelabs/curator/request_processor/base_online_request_processor.py:483]: get_rate_limits is defined as an abstract method base_request_processor.py
For batches, there are rate limits TPD, but we just aren't doing jobs that large (yet).

Yes you understand correctly and I'll implement this change - thanks for the suggestion@

File: src/bespokelabs/curator/request_processor/litellm_online_request_processor.py [hunk 2]
```diff
@@ -134,6 +138,24 @@ def estimate_total_tokens(self, messages: list) -> int:
...
+        try:
+            litellm.completion_cost(completion_response=completion.model_dump())
+        except litellm.NotFoundError as e:
+            logger.warning(f"LiteLLM does not support cost estimation for model {self.model}: {e}")
+
+        headers = completion._hidden_params.get("additional_headers", {})
+        logger.info(f"Test call headers: {headers}")
+        return headers
+
     def get_rate_limits(self) -> dict:
```
[the remark this answers is not in the document]
Comment (RyanMarten) [reply src/bespokelabs/curator/request_processor/litellm_online_request_processor.py:159]: yes

**Review by vutrung96 — APPROVED** [review]
LGTM!

**Pull request merged**